examples/pxScene2d/src/pxShaderResource.cpp

```diff
@@ -0,0 +1,1389 @@
+  {
+    return pxShaderResource::setUniforms( value->toObject() );
+  }
+  if(strcmp(name, "t") == 0)
+  {
+    return RT_OK; // ignore
+  }
+  else
+  {
+    return setUniformVal(name, *value);
+  }
+}
+
+
+//
+// Stores the UNIFORMS parsed from JS code/config; to be located in Shader when compiled
+//
+rtError pxShaderResource::uniforms(rtObjectRef& /* o */) const { return RT_OK; }
+rtError pxShaderResource::setUniforms(rtObjectRef o)
+{
+  rtValue allKeys;
+  o->Get("allKeys", &allKeys);
+
+  rtArrayObject* keys = (rtArrayObject*) allKeys.toObject().getPtr();
+  for (uint32_t i = 0, l = keys->length(); i < l; ++i)
+  {
+    rtValue keyVal;
+    if (keys->Get(i, &keyVal) == RT_OK && !keyVal.isEmpty())
+    {
+      setFunc_t     setFunc = NULL;
+      UniformType_t typeEnum = UniformType_Unknown;
+
+      rtString name = keyVal.toString();
+      rtString type = o.get<rtString>(name);
+
+      if(name == "u_time")
+      {
+        mIsRealTime = true;
+      }
+      else
+      if(type == "int")
+      {
+        setFunc  = pxShaderResource::setUniform1i;
+        typeEnum = UniformType_Int;
+      }
+      else
+      if(type == "float")
+      {
+        setFunc  = pxShaderResource::setUniform1f;
+        typeEnum = UniformType_Float;
+      }
+      else
+      if(type == "ivec2")
+      {
+        setFunc  = pxShaderResource::setUniform2iv;
+        typeEnum = UniformType_iVec2;
+      }
+      else
+      if(type == "ivec3")
+      {
+        setFunc  = pxShaderResource::setUniform3iv;
+        typeEnum = UniformType_iVec3;
+      }
+      else
+      if(type == "ivec4")
+      {
+        setFunc  = pxShaderResource::setUniform4iv;
+        typeEnum = UniformType_iVec4;
+      }
+      else
+      if(type == "vec2")
+      {
+        setFunc  = pxShaderResource::setUniform2fv;
+        typeEnum = UniformType_Vec2;
+      }
+      else
+      if(type == "vec3")
+      {
+        setFunc  = pxShaderResource::setUniform3fv;
+        typeEnum = UniformType_Vec3;
+      }
+      else
+      if(type == "vec4")
+      {
+        setFunc  = pxShaderResource::setUniform4fv;
+        typeEnum = UniformType_Vec4;
+      }
+      else
+      if(type == "sampler2D")
+      {
+        if(name == "s_texture")
+        {
+          mNeedsFbo = true;
+        }
+        setFunc = (mSamplerCount == 3) ? pxShaderResource::bindTexture3 :
+                  (mSamplerCount == 4) ? pxShaderResource::bindTexture4 :
+                  (mSamplerCount == 5) ? pxShaderResource::bindTexture5 : NULL;
+
+        typeEnum = UniformType_Sampler2D;
+
+        mSamplerCount++;
+      }
+
+      mUniform_map[name] = { name, typeEnum, (GLint) -1, rtValue(), false, setFunc };  // ADD TO MAP
+    }
+  }//FOR
+
+  return RT_OK;
+}
+
+void pxShaderResource::loadResourceFromFile()
+{
+  init();
+
+  // Since this object can be released before we get a async completion
+  // We need to maintain this object's lifetime
+  // TODO review overall flow and organization
+  AddRef();
```

Comment:
There is an extra `AddRef()` call in `loadResourceFromFile()`.  The `AddRef()` calls before the calls to `gUIThreadQueue->addTask()` look good, but it appears that we're calling `AddRef()` one too many times in `pxShaderResource::loadResourceFromFile()`.  This will result in a resource leak when loading shaders from a file.

Reply:
Thanks for spotting this.  Removed this call.